### ISSUE TITLE
Elixir 1.3 is periodically losing protocol implementations

### ISSUE THREAD

**Issue by myronmarston** [elixir-lang/elixir#4842]
### Environment

* Elixir version (elixir -v):

```
Erlang/OTP 18 [erts-7.3] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Elixir 1.3.0
```

* Operating system:

OS X El Capitan

### Current behavior

Several times today, protocol implementations have vanished, and I start getting errors like:

```
** (Protocol.UndefinedError) protocol Delorean.Shard not implemented for %Delorean.CrawlModels.CrawlShard{campaign_id: "prod.1.10", entity_dicts: %{urls: %Delorean.Util.ValueDict{data: {}}}, in_memory_data: %Delorean.Shard.InMemoryData{builders_ref: #Reference<0.0.3.6503>, memory_usage: 592, owner: nil, shard_ref: #Reference<0.0.3.6502>, usage_mode: :default}, metadata: %{build_history: [], snapshots_by_backend: %{}}, page_issues: [], version: {0, 1}}
```

If I `rm -rf _build` and recompile, the problem goes away.

Unfortunately, I have not yet been able to figure out a consistent set of steps that triggers this, but I think it has to do with the fact we have an umbrella app and sometimes I run `mix test` in the umbrella root and sometimes I run `mix test` from within one of the sub-apps (or rather, my vim keybinding does that for me 😄 ).  I usually notice the error on one of the subsequent test runs _after_ I've run a mix task from one of the sub-apps, so it appears to be losing the protocol implementations at that point.

### Expected behavior

Protocol implementations should not be lost.  Once they are compiled they should remain available for subsequent runs of `mix test` (or any other mix task).

**Comment by michalmuskala** [comment]
Do you have `build_path` set up in the applications? And do you have proper dependencies between the applications?

**Comment by myronmarston** [comment]
> Do you have `build_path` set up in the applications?

Yep.  We've got this for each:

``` elixir
# ...
deps_path: "../../deps",
lockfile: "../../mix.lock",
build_path: "../../_build",
config_path: "../../config/config.exs",
# ...
```

> And do you have proper dependencies between the applications?

Yes, we have dependencies declared between the applications.  It's possible we've missed some declarations, of course, so it may not be "proper".

I realize I didn't give much to go off of to reproduce this bug (sorry about that!).  I plan to spend some time in the next day or two trying to come up with an isolated repro case.

**Comment by josevalim** [comment]
@myronmarston thank you for the report, we are looking towards more news.

Btw, did this happen in the RCs? We did change parts of this code to make compilation in umbrella faster so we may have introduced bugs BUT we didn't really change it after RCs. So it would be interesting to see what happened there.

**Comment by myronmarston** [comment]
> Btw, did this happen in the RCs?

Not that I noticed, but I didn't actually try the RCs for any extended period of time.  I was still using 1.2 for my normal development while using the RCs to get my app warning-free and green on 1.3.  I'm not sure if I ever followed whatever sequence triggers this while using the RCs.

Your question gives me a good idea, though: I'm going to use `git bisect` and see if I can isolate this regression to a particular commit.

**Comment by myronmarston** [comment]
I've run `git bisect` to find the source of this regression.  It's cd9f4e1cf0dc4032bae538b22407a597e71c871a.  On that commit I see the protocol implementations get lost; on the parent commit, fd8230f25a42499398b6c28127d6645d473debe4, the problem does not occur.

Here is the sequence of commands I'm running from within my repository to reproduce this:

``` bash
# ensure we have no existing build artifacts
$ rm -rf _build
# run the whole test suite
$ mix test 
# run the test suite of one of the sub apps (`rankings_endpoint_models`)
$ (cd apps/rankings_endpoint_models && mix test)
# Modify one of the source files of the sub app to force it to build again
$ echo "" >> apps/rankings_endpoint_models/lib/serp_features.ex
# Run one test file in the sub app (which also causes the file edited above to re-build)
$ (cd apps/rankings_endpoint_models && mix test test/unit/serp_features_test.exs) 
# Run the entire tes suite again
$ mix test
```

When this last `mix test` runs, the protocol implementations are not loaded and I get the `Protocol.UndefinedError`.

BTW, this confirms that this was happening with the RCs--I just hadn't run into it since I hadn't run this sequence of commands.

@josevalim -- is that enough for you to go off of?

**Comment by josevalim** [comment]
@myronmarston I will investigate. :D Thank you!

**Comment by myronmarston** [comment]
BTW, when this happens I've noticed that my test suite also takes much longer to run -- usually 2-3x longer.  I'm guessing that the missing protocol implementations have something to do with it?

**Comment by josevalim** [comment]
@myronmarston is the protocol defined in a dependency or in another umbrella app?

**Comment by myronmarston** [comment]
> @myronmarston is the protocol defined in a dependency or in another umbrella app?

Another umbrella app.

**Comment by josevalim** [comment]
@myronmarston and that's not the umbrella app you are touching right?

**Comment by josevalim** [comment]
@myronmarston I need a bit more information:

1. Which app defines the protocol being lost
2. How the application that is being touched above relates to the one that defines the protocol. Is it a direct in_umbrella dependency? Or is it a dep of a dep?

**Comment by myronmarston** [comment]
> Which app defines the protocol being lost

The protocol is `Delorean.Shard` and is defined in `apps/shard`.

> How the application that is being touched above relates to the one that defines the protocol. Is it a direct in_umbrella dependency? Or is it a dep of a dep?

Here's the output from `mix deps.tree` (with a bunch of `--exclude` flags so we don't include things like ecto, timex and mariaex):

```
rankings_endpoint_models
├── csv ~> 1.2 (Hex package)
├── common_models (../common_models)
├── rankings_models (../rankings_models)
│   └── common_models (../common_models)
├── rankings_shard_repository (../rankings_shard_repository)
│   ├── common_models (../common_models)
│   ├── rankings_models (../rankings_models)
│   ├── shard_storage (../shard_storage)
│   │   ├── common_models (../common_models)
│   │   ├── runtime_config (../runtime_config)
│   │   │   └── ex_json_schema ~> 0.3 (Hex package)
│   │   └── shard (../shard)
│   ├── statsd (../statsd)
│   │   ├── runtime_config (../runtime_config)
│   │   └── ex_statsd >= 0.5.3 (Hex package)
│   └── benchfella ~> 0.2 (Hex package)
└── rankings_builder (../rankings_builder)
    ├── campaign_status (../campaign_status)
    │   ├── common_models (../common_models)
    │   └── runtime_config (../runtime_config)
    ├── common_models (../common_models)
    ├── packrat_snapshot_repository (../packrat_snapshot_repository)
    │   ├── common_models (../common_models)
    │   └── shard (../shard)
    ├── rankings_models (../rankings_models)
    ├── rankings_shard_repository (../rankings_shard_repository)
    ├── statsd (../statsd)
    └── ex_json_schema ~> 0.3 (Hex package)
```

As you can see, there are a couple dependency paths:

* `rankings_endpoint_models` -> `rankings_shard_repository` -> `shard_storage` -> `shard`
* `rankings_endpoint_mdoels` -> `rankings_builder` -> `packrat_snapshot_repository` -> `shard`

**Comment by myronmarston** [comment]
Thanks for the fix, @josevalim!  I can confirm it fixes the problem for me.

I was going to ask if this could be included in 1.3.1 but I see that's already been released.  For now my team will probably have to compile against HEAD ourselves to avoid this issue.  Do you plan to include this in 1.3.2?

**Comment by josevalim** [comment]
> For now my team will probably have to compile against HEAD ourselves to avoid this issue.

I don't understand. This bug reveals a bug in your code as the :crawl_models was not a dependency of rankings_endpoint_models. So as long as you fix this bug, this patch should be completely irrelevant to you. So why do you need HEAD? :)

I don't plan to ship this as part of 1.3.2 because this area of the code, tied to umbrellas, is complex, so I would prefer to battle test it for longer than backport to the stable branch, given this patch only affects your code when there is a bug (i.e. a missing application).

**Comment by myronmarston** [comment]
> I don't understand. This bug reveals a bug in your code as the :crawl_models was not a dependency of rankings_endpoint_models. So as long as you fix this bug, this patch should be completely irrelevant to you. So why do you need HEAD? :)

I don't understand what the bug in our code was.   `crawl_models` is indeed not a dependency of `rankings_endpoint_models` and should not be.  I tested making it dependency to see if making it a dependency worked around this issue and it works...but there is nothing in `rankings_endpoint_models` that has anything to do with `crawl_models`, so it is definitely wrong for us to declare it as a dependency when it does not depend on it in anyway.

**Comment by josevalim** [comment]
Oh, sorry, you are right. I misunderstood part of the issue: the test that fails is not part of `rankings_endpoint_models`. I will backport the issue to v1.3.2 and I'd appreciate if you give the v1.3 branch a try for a while. THank you!

**Comment by myronmarston** [comment]
> Oh, sorry, you are right. I misunderstood part of the issue: the test that fails is not part of rankings_endpoint_models. I will backport the issue to v1.3.2 and I'd appreciate if you give the v1.3 branch a try for a while. THank you!

Thanks!  I've been trying it since yesterday.  I've been periodically seeing a new failure I haven't seen before that seems to be related.  Occasionally when I run `mix test path/to/test.exs` from within on of our sub-apps, I get:

```
** (EXIT from #PID<0.46.0>) an exception was raised:
    ** (MatchError) no match of right hand side value: {:error, :no_beam_info}
        (mix) lib/mix/tasks/compile.protocols.ex:125: Mix.Tasks.Compile.Protocols.consolidate/4
        (elixir) lib/task/supervised.ex:94: Task.Supervised.do_apply/2
        (elixir) lib/task/supervised.ex:45: Task.Supervised.reply/5
        (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

11:33:07.537 [error] Task #PID<0.467.0> started from #PID<0.46.0> terminating
** (MatchError) no match of right hand side value: {:error, :no_beam_info}
    (mix) lib/mix/tasks/compile.protocols.ex:125: Mix.Tasks.Compile.Protocols.consolidate/4
    (elixir) lib/task/supervised.ex:94: Task.Supervised.do_apply/2
    (elixir) lib/task/supervised.ex:45: Task.Supervised.reply/5
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: #Function<8.112345637/0 in Mix.Tasks.Compile.Protocols.consolidate/6>
    Args: []
```

It goes away when I retry.  I don't think I saw this before running against an Elixir version with your fix applied, so it might be related.  Let me know if you want me to open a new issue with this info.

**Comment by josevalim** [comment]
Does it happen when running on the umbrella, inside a child app, or both?

-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by myronmarston** [comment]
It happens when running inside a child app.  I have not noticed it when running in an umbrella, but I run tasks from within a child app much more often than running tasks from the umbrella root (my main workflow involves a vim keybinding that runs a specific test file by invoking `mix test path/to/file.exs` from within its app) -- so it's possible I just haven't run into it in the umbrella yet.

**Comment by myronmarston** [comment]
I figured out a command that consistently triggers it:

```
➜  campaign_status git:(myron/rebuild-stale) ✗ touch lib/build_management/build_management.ex; mix test test/integration/build_management_test.exs; mix test test/integration/build_management_test.exs
Compiling 1 file (.ex)
** (EXIT from #PID<0.46.0>) an exception was raised:
    ** (MatchError) no match of right hand side value: {:error, :no_beam_info}
        (mix) lib/mix/tasks/compile.protocols.ex:125: Mix.Tasks.Compile.Protocols.consolidate/4
        (elixir) lib/task/supervised.ex:94: Task.Supervised.do_apply/2
        (elixir) lib/task/supervised.ex:45: Task.Supervised.reply/5
        (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

11:50:14.757 [error] Task #PID<0.454.0> started from #PID<0.46.0> terminating
** (MatchError) no match of right hand side value: {:error, :no_beam_info}
    (mix) lib/mix/tasks/compile.protocols.ex:125: Mix.Tasks.Compile.Protocols.consolidate/4
    (elixir) lib/task/supervised.ex:94: Task.Supervised.do_apply/2
    (elixir) lib/task/supervised.ex:45: Task.Supervised.reply/5
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: #Function<8.112345637/0 in Mix.Tasks.Compile.Protocols.consolidate/6>
    Args: []
Excluding tags: [:manual]

................................

Finished in 0.5 seconds
32 tests, 0 failures

Randomized with seed 582130
```

Running `mix test path/to/test.exs` after saving an implementation file consistently triggers the error (apparently because it had to recompile that file).  But if I just run `mix test path/to/test.exs` (or run it after the failure) it works.

**Comment by josevalim** [comment]
path/to/test.exs can be any file, right? Does this happen in any app as
well?

On Wednesday, June 29, 2016, Myron Marston <notifications@github.com> wrote:

> I figured out a command that consistently triggers it:
>
> ➜  campaign_status git:(myron/rebuild-stale) ✗ touch lib/build_management/build_management.ex; mix test test/integration/build_management_test.exs; mix test test/integration/build_management_test.exs
> Compiling 1 file (.ex)
> ** (EXIT from #PID<0.46.0>) an exception was raised:
>     ** (MatchError) no match of right hand side value: {:error, :no_beam_info}
>         (mix) lib/mix/tasks/compile.protocols.ex:125: Mix.Tasks.Compile.Protocols.consolidate/4
>         (elixir) lib/task/supervised.ex:94: Task.Supervised.do_apply/2
>         (elixir) lib/task/supervised.ex:45: Task.Supervised.reply/5
>         (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
>
> 11:50:14.757 [error] Task #PID<0.454.0> started from #PID<0.46.0> terminating
> ** (MatchError) no match of right hand side value: {:error, :no_beam_info}
>     (mix) lib/mix/tasks/compile.protocols.ex:125: Mix.Tasks.Compile.Protocols.consolidate/4
>     (elixir) lib/task/supervised.ex:94: Task.Supervised.do_apply/2
>     (elixir) lib/task/supervised.ex:45: Task.Supervised.reply/5
>     (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
> Function: #Function<8.112345637/0 in Mix.Tasks.Compile.Protocols.consolidate/6>
>     Args: []
> Excluding tags: [:manual]
>
> ................................
>
> Finished in 0.5 seconds
> 32 tests, 0 failures
>
> Randomized with seed 582130
>
> Running mix test path/to/test.exs after saving an implementation file
> consistently triggers the error (apparently because it had to recompile
> that file). But if I just run mix test path/to/test.exs (or run it after
> the failure) it works.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/issues/4842#issuecomment-229452589>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAlbm8q3FTispsAdZ107GMiGaCLnGDbks5qQr7TgaJpZM4I7SnR>
> .
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by myronmarston** [comment]
> path/to/test.exs can be any file, right?

Any test file.

> Does this happen in any app as well?

It appears to.  I haven't tried it in all of our sub apps (we have about 40 of them!) but I've tried it in several and it's happened in all of them so far.

**Comment by myronmarston** [comment]
Hmm, it is apparently not happening of the sub apps.  I just found one (named `rankings_builder`) for which it does not occur.  I'm not yet sure what's different about it (any suggestions of what to look into?).

**Comment by myronmarston** [comment]
I'm also running into a case where re-running `mix test path/to/file.exs` does not fix the issue.  The failure keeps happening every time I try to run a test and I have to blow about my `_build` directory to fix it.

**Comment by josevalim** [comment]
@myronmarston can you please do this change in your Elixir checkout:

```diff
diff --git a/lib/mix/lib/mix/tasks/compile.protocols.ex b/lib/mix/lib/mix/tasks/compile.protocols.ex
index 8713d5d..e2632b9 100644
--- a/lib/mix/lib/mix/tasks/compile.protocols.ex
+++ b/lib/mix/lib/mix/tasks/compile.protocols.ex
@@ -120,13 +120,13 @@ defmodule Mix.Tasks.Compile.Protocols do
   end

   defp consolidate(protocol, paths, output, opts) do
+    if opts[:verbose] do
+      Mix.shell.info "Consolidating #{inspect protocol}"
+    end
     impls = Protocol.extract_impls(protocol, paths)
     reload(protocol)
     {:ok, binary} = Protocol.consolidate(protocol, impls)
     File.write!(Path.join(output, "#{protocol}.beam"), binary)
-    if opts[:verbose] do
-      Mix.shell.info "Consolidated #{inspect protocol}"
-    end
   end

   defp reload(module) do
```

and then run `make mix` to compile it and reproduce the error so we know which protocol is failing? I need to know if it is a protocol defined in the current project, a dependency or in Elixir. :) I will push this change to elixir permanently as well so we get better reports in cases like this.

**Comment by josevalim** [comment]
@myronmarston or, instead of running the changes above, just let me know which implementations are defined inside "lib/build_management/build_management.ex". :)

**Comment by myronmarston** [comment]
> @myronmarston or, instead of running the changes above, just let me know which protocols and implementations are defined inside "lib/build_management/build_management.ex". :)

As it turns out, nothing is defined in that file.  It wasn't an existing file (the file I was thinking of was `lib/build_management.ex`) but `touch`ing it created an empty file.

Anyhow, I applied the diff above and now I'm getting:

```
➜  campaign_status git:(myron/travis-build-improvements) MIX_ENV=test mix compile --verbose
Compiling 2 files (.ex)
Compiled test/support/build_management_stub.ex
Compiled lib/build_management.ex
Consolidating Delorean.Shard
Consolidating Inspect
** (EXIT from #PID<0.46.0>) an exception was raised:
    ** (MatchError) no match of right hand side value: {:error, :no_beam_info}
        (mix) lib/mix/tasks/compile.protocols.ex:128: Mix.Tasks.Compile.Protocols.consolidate/4
        (elixir) lib/task/supervised.ex:94: Task.Supervised.do_apply/2
        (elixir) lib/task/supervised.ex:45: Task.Supervised.reply/5
        (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

11:34:01.056 [error] Task #PID<0.467.0> started from #PID<0.46.0> terminating
** (MatchError) no match of right hand side value: {:error, :no_beam_info}
    (mix) lib/mix/tasks/compile.protocols.ex:128: Mix.Tasks.Compile.Protocols.consolidate/4
    (elixir) lib/task/supervised.ex:94: Task.Supervised.do_apply/2
    (elixir) lib/task/supervised.ex:45: Task.Supervised.reply/5
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: #Function<8.98043743/0 in Mix.Tasks.Compile.Protocols.consolidate/6>
    Args: []
```

Apparently the error is happening while consolidating the `Inspect` protocol, but I don't have any implementations of it in this sub-app.  I do in a couple of the sibling apps, though.

**Comment by myronmarston** [comment]
Just now I ran into the `Protocol.UndefinedError` error I originally reported in this issue, even though I'm running against 301aa6eddce60c02a7fd11a7c1426aec97338878 which has your fix for the issue.  I'm not sure why.  It's possible that I accidentally ran a mix task in another shell which doesn't have my locally compiled elixir on the path but I'm 99% sure I didn't do that.  So it's a bit of a mystery.  I'll let you know if I find out more.

**Comment by wojtekmach** [comment]
I think I'm running into similar issue, I'm on `Elixir 1.4.0-dev (1498c42)`

Here's a simplified repo to reproduce: https://github.com/wojtekmach/umbrella_bug:
- [apps/money/lib/money.ex](https://github.com/wojtekmach/umbrella_bug/blob/master/apps/money/lib/money.ex) defines a `Money` struct
- [apps/web/lib/money_ext.ex](https://github.com/wojtekmach/umbrella_bug/blob/master/apps/web/lib/money_ext.ex) defines a `Phoenix.HTML.Safe` implementation for Money [1]
- I added `:phoenix_html` `applications: ` in `app/web/mix.exs` - not sure if I needed to. Decided to just depend on `phoenix_html` instead of pulling entire `phoenix.

I can reliably reproduce this by running `cd apps/money && touch lib/money.ex && MIX_ENV=test mix compile --verbose`, I'm getting: 

```
Compiling 1 file (.ex)
Compiled lib/money.ex
** (EXIT from #PID<0.46.0>) an exception was raised:
    ** (MatchError) no match of right hand side value: {:error, :no_beam_info}
        (mix) lib/mix/tasks/compile.protocols.ex:125: Mix.Tasks.Compile.Protocols.consolidate/4
        (elixir) lib/task/supervised.ex:94: Task.Supervised.do_apply/2
        (elixir) lib/task/supervised.ex:45: Task.Supervised.reply/5
        (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

18:37:11.197 [error] Task #PID<0.78.0> started from #PID<0.46.0> terminating
** (MatchError) no match of right hand side value: {:error, :no_beam_info}
    (mix) lib/mix/tasks/compile.protocols.ex:125: Mix.Tasks.Compile.Protocols.consolidate/4
    (elixir) lib/task/supervised.ex:94: Task.Supervised.do_apply/2
    (elixir) lib/task/supervised.ex:45: Task.Supervised.reply/5
    (stdlib) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: #Function<8.71273319/0 in Mix.Tasks.Compile.Protocols.consolidate/6>
    Args: []
```

Hope this is useful, let me know if I can help in any way. /cc @josevalim @myronmarston 

[1] I didn't want to put `Phoenix.HTML.Safe` impl in `money` app because I didn't want to add phoenix dep. Guarding with `if Code.ensure_loaded?(Phoenix.HTML.Safe)` do: defimpl ...` inside `apps/money/lib/money.ex` made the compilation flaky - if I did: `cd apps/money && mix test` it was OK, but then `cd apps/web && mix test` was complaining about missing protocol.

**Comment by josevalim** [comment]
@wojtekmach beautiful, I will investigate it now. Btw, huge thank you to both you and @myronmarston for using development versions and helping us find bugs early!

**Comment by josevalim** [comment]
Fixed in master and v1.3. Please @myronmarston continue giving it a try and let me know if anything new shows up.

**Comment by myronmarston** [comment]
Thank you @josevalim!

> Please @myronmarston continue giving it a try and let me know if anything new shows up.

Will do!
